astropy/cosmology/parameter.py

```diff
@@ -317,13 +328,19 @@ def _validate_with_unit(cosmology, param, value):
     Default Parameter value validator.
     Adds/converts units if Parameter has a unit.
     """
-    if param.unit is not None:
-        with u.add_enabled_equivalencies(param.equivalencies):
-            value = u.Quantity(value, param.unit)
+    # Quantity with right units (can be None) and dtype
+    with u.add_enabled_equivalencies(param.equivalencies):
```

Comment:
This really seems to be the core of the change, which I like. But do we really want to use non-float for `Quantity`?

Reply:
The non-float is a nice flexibility, but I don't anticipate it being useful for a while. Maybe something with discrete cosmologies or something.
I think the more powerful use case is controlling the precision of floats. E.g. float128 over float64. Or degrading precision for faster speed. Currently the ``dtype`` is a class-level setting, but I hope to build up the machinery for setting the precision at instantiation.

Reply:
I guess one way to do that is to have ``dtype=any float-like`` and then use the dtype passed at instantiation, checking that it's floatlike.

Reply:
In which case... the float-like check is a validator! 
I still think ``_validate_to_float`` should be deprecated.
In future I'll probably add

```
@Parameter.register_validator("float-like")
def _validate_is_floating(cosmology, param, value):
    """Parameter value validator with units, and checked for a  floating dtype."""
    value = _validate_with_unit(cosmology, param, value)
    if not isinstance(getattr(value, "value", value), (float, np.floating)):  # check dtype, degrading Quantity to ndarray
        raise TypeError(...)
    return value
```

Then the default ``dtype`` can be None. Quantity can upcast things that are meant to be floats and they will pass ``_validate_is_floating``. For things not meant to be floats, the Parameter.dtype can be specified!

Reply:
Or maybe this worth doing now...

Reply:
This does also re-raise your original question if adding ``dtype`` to ``Parameter`` is the best way to go about doing this. The ``_validate_is_floating `` may be the better solution.

Reply:
OK, the use of `power` then definitely is OK.

I do think that it would be best to only introduce `dtype` when you actually need it...

Reply:
I might actually close this PR then and just do a PR about changing ``float`` to ``float64``.

Reply:
Or never mind. All I have to do is get rid of a few lines about ``dtype``.

Reply:
Hmm, that might indeed be best. It may be getting to the stage where you want to write a trial multi-d cosmology.  lots of thought needed, e.g., on how to do broadcasting. Might be good to start with some concrete use cases.

Reply:
My current thought is that the shape will be ``(*parameters_shape, *redshift_array_shape)`` for methods that take redshift arrays. So 

``cosmo.hubble_time`` will have shape ``(*parameters_shape)`` while 
``cosmo.age([1, 2, 3])`` will have shape ``(*parameters_shape, 3)``.

Reply:
You'll need to be a bit careful here on how you want different calculations to broadcast. E.g., with that scheme, the different cosmologies will not broadcast properly if you also calculate a scalar time difference and want to add that (not sure that would be a good idea...).

It might be better to go the other way, though my overall guess is that it is in fact best to just broadcast the cosmology with the input, i.e., in your example it would be up to the user to do `cosmo[:, np.newaxis].age([1, 2, 3])` or `cosmo.age([[1], [2], [3]])`.  I think that in the end gives most flexibility. It may also give you the easiest route to dealing with different parameters having different shapes. 

p.s.  Another alternative would be to be a bit more like `Distribution` and use a structured array for the cosmology. But that might prevent using `Distribution` as parameters.

p.p.s. May be good to take this out to a separate issue!

Reply:
p.p.p.s. If you haven't done so already, have a look at `utils.shapes.ShapedLikeNDArray`, which is used inside `Time` and `SkyCoord`, etc.

Reply:
Great! Thanks for all the suggestions!

Reply:
@mhvk 

> p.s. Another alternative would be to be a bit more like Distribution and use a structured array for the cosmology. But that might prevent using Distribution as parameters.

I was thinking along these same lines: of having a `_parameters` structured Quantity and all the parameters, eg `.H0`, be views.

But I very much want parameters to be Distributions! Actually maybe I want the whole cosmology to be a distribution, like Quantity. Have to think on that.

Reply:
I *think* you'd want the parameters to be `Distribution`, since in typical use some surely only some of them would be sampled, while others would not be. Also, `Distribution` very much only works with `ndarray` subclasses so far.